library/L1_Peripheral/lpc40xx/eeprom.hpp

```diff
@@ -0,0 +1,155 @@
+  static constexpr sjsu::lpc40xx::SystemController kLpc40xxSystemController =
+      sjsu::lpc40xx::SystemController();
+
+  // Constructor has an optional SystemController parameter which can be
+  // specified for testing purposes. Otherwise, the actual SystemController
+  // object is used (kLpc40xxSystemController)
+  explicit constexpr Eeprom(const sjsu::SystemController & system_controller =
+                             kLpc40xxSystemController)
+      : system_controller_(system_controller)
+  {
+  }
+
+  // Initializing the EEPROM requires setting the wait state register,
+  // and the clock divider register
+  void Initialize() const override
+  {
+    const float kSystemClock    = static_cast<float>
+                                    (system_controller_.GetSystemFrequency());
+    constexpr float kEepromClk  = 375'000;  // EEPROM runs at 375 kHz
+    constexpr float kNanosecond = 1E-9f;
+
+    // Initialzie EEPROM wait state register with number of wait states
+    // for each of its internal phases
+    // Phase 3 (15 ns)
+    eeprom_register->WSTATE |=
+        static_cast<uint8_t>((15 * kNanosecond * kSystemClock) + 1);
+    // Phase 2 (55 ns)
+    eeprom_register->WSTATE |=
+        (static_cast<uint8_t>((55 * kNanosecond * kSystemClock) + 1)) << 8;
+    // Phase 1 (35 ns)
+    eeprom_register->WSTATE |=
+        (static_cast<uint8_t>((35 * kNanosecond * kSystemClock) + 1)) << 16;
+
+    // Initialize EEPROM clock
+    eeprom_register->CLKDIV = static_cast<uint8_t>(kSystemClock / kEepromClk);
+  }
+
+  // The EEPROM can only be interfaced in 32-bit mode, which means that the
+  // address used must be a multiple of 4. The address entered will be cut
+  // down to a multiple of 4.
+  void Write(const uint8_t * wdata, uint32_t full_address,
+             size_t count) const override
+  {
+    constexpr uint8_t kMax6Bits = 0b11'1111;
+
+    // Writing a 32-bit number to an address that isn't a multiple of 4 will
+    // cause a hard fault, thus this safety is put into place to avoid that.
+    full_address &= 0b1111'1111'1100;
+
+    // Because the EEPROM uses 32-bit communication, write_data will be casted
+    // into a uint32_t *
+    const uint32_t * write_data = reinterpret_cast<const uint32_t *>(wdata);
+
+    // The first 6 bits in the address (MSB) dictate which page is being written
+    // to in the EEPROM, and the last 6 bits (LSB) dictate offset in the page
+    uint8_t page_count  = static_cast<uint8_t>(full_address >> 6);
+    uint8_t page_offset = full_address & 0b0000'0011'1111;
+
+    uint16_t address;
+
+    for (; count > 0; count -= 4)
+    {
+      // Page offset is incremented by 4 because we're writing 32 bits
+      constexpr uint8_t kOffsetInterval = 4;
+
+      address = static_cast<uint16_t>((page_count << 6) + page_offset);
+
+      eeprom_register->ADDR  = address;
+      eeprom_register->CMD   = kWrite32Bits;
+      eeprom_register->WDATA = write_data[0];
+
+      write_data++;

```

Comment:
Refrain from doing C style pointer arithmetic. This can be easily solved by using bracket notation:

`write_data[i]` rather than `write_data[0]` each time.